### PR TITLE
[EC] Unify point doubling for P-256/384/521

### PR DESCRIPTION
**Pull request by dkostic**
### Issues:
CryptoAlg-2408

### Description of changes: 
Implement and use a single version of point doubling
for implementations of NIST curves P-384, P-521, and
Fiat-crypto based implementation of P-256. The change
does not affect performance.

Point addition will be unified in a subsequent change.

### Call-outs:
I verified the performance was not affected on Graviton 3, Intel, and M1 CPUs. Example for M1:
```
Before
Did 2882750 EC POINT P-384 dbl operations in 1000082us (2882513.6 ops/sec)
Did 1600000 EC POINT P-384 add operations in 1000497us (1599205.2 ops/sec)
Did 7051 EC POINT P-384 mul operations in 1078289us (6539.1 ops/sec)
Did 28000 EC POINT P-384 mul base operations in 1000115us (27996.8 ops/sec)
Did 5632 EC POINT P-384 mul public operations in 1062456us (5300.9 ops/sec)
Did 2685500 EC POINT P-521 dbl operations in 1000037us (2685400.6 ops/sec)
Did 1435000 EC POINT P-521 add operations in 1000129us (1434814.9 ops/sec)
Did 4928 EC POINT P-521 mul operations in 1055318us (4669.7 ops/sec)
Did 19000 EC POINT P-521 mul base operations in 1022199us (18587.4 ops/sec)
Did 3850 EC POINT P-521 mul public operations in 1036809us (3713.3 ops/sec)

After:
Did 2888250 EC POINT P-384 dbl operations in 1000028us (2888169.1 ops/sec)
Did 1593000 EC POINT P-384 add operations in 1000405us (1592355.1 ops/sec)
Did 6875 EC POINT P-384 mul operations in 1054301us (6520.9 ops/sec)
Did 28000 EC POINT P-384 mul base operations in 1000818us (27977.1 ops/sec)
Did 5555 EC POINT P-384 mul public operations in 1056370us (5258.6 ops/sec)
Did 2775250 EC POINT P-521 dbl operations in 1000021us (2775191.7 ops/sec)
Did 1435000 EC POINT P-521 add operations in 1000085us (1434878.0 ops/sec)
Did 4840 EC POINT P-521 mul operations in 1044164us (4635.3 ops/sec)
Did 19000 EC POINT P-521 mul base operations in 1027887us (18484.5 ops/sec)
Did 3883 EC POINT P-521 mul public operations in 1051447us (3693.0 ops/sec)
```

### Testing:
How is this change tested (unit tests, fuzz tests, etc.)? Are there any testing steps to be verified by the reviewer?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

**Review by nebeid — DISMISSED** [review]

File: crypto/fipsmodule/ec/ec_nistp.c [hunk 1]
```diff
@@ -0,0 +1,111 @@
+/*
```
Comment (torben-hansen) [inline crypto/fipsmodule/ec/ec_nistp.c:1]: Should be this:
```
// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0 OR ISC
```

File: crypto/fipsmodule/ec/ec_nistp.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
+/*
```
Comment (torben-hansen) [inline crypto/fipsmodule/ec/ec_nistp.h:1]: Same comment as above

File: crypto/fipsmodule/ec/ec_nistp.c [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+//   2. point doubling,
+//   3. scalar multiplication of a base point,
+//   4. scalar multiplication of an arbitrary point,
+//   5. scalar multiplication of a base and an arbitrary point.
+//
+// Matrix of what has been done so far:
+// 
+// | op | P-521 | P-384 | P-256 |
+// |----------------------------|
+// | 1. |   x   |   x   |   x*  |
```
Comment (torben-hansen) [inline crypto/fipsmodule/ec/ec_nistp.c:22]: I believe you are implementing doubling not addition below.

File: crypto/fipsmodule/ec/ec_nistp.c [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+  // x' = alpha^2 - 8*beta
+  ctx->sqr(x_out, alpha);
+  ctx->add(fourbeta, beta, beta);
+  ctx->add(fourbeta, fourbeta, fourbeta);
+  ctx->add(tmptmp, fourbeta, fourbeta);
+  ctx->sub(x_out, x_out, tmptmp);
+
+  // z' = (y + z)^2 - gamma - delta
+  // The following calculation differs from the Coq proof cited above.
```
Comment (torben-hansen) [inline crypto/fipsmodule/ec/ec_nistp.c:91]: okay but then we can't cite the Coq proof? Because it proves something different.

Comment (torben-hansen) [reply crypto/fipsmodule/ec/ec_nistp.c:91]: Or is there another proof for this variant?

Comment (dkostic) [reply crypto/fipsmodule/ec/ec_nistp.c:92]: you are right, I clarified this in the code comment, let me know if that's fine with you.

**Review by torben-hansen — APPROVED** [review]

**Review by nebeid — APPROVED** [review]

**Pull request merged**